File: 1447-jump-game-iv/jump-game-iv.py
class Solution:
    def minJumps(self, nums: List[int]) -> int:
        idx_map = collections.defaultdict(list)
        for i, num in enumerate(nums):
            idx_map[num].append(i)

        # bfs        
        visited = set([0])
        q = collections.deque([0])
        min_jumps = 0

        while q:
            level_size = len(q)
            for _ in range(level_size):
                i = q.popleft()
                if i == len(nums) - 1:
                    return min_jumps

                for nbr in idx_map[nums[i]] + [i - 1, i + 1]:
                    if nbr not in visited and nbr >= 0 and nbr < len(nums):
                        q.append(nbr)
                        visited.add(nbr)
                idx_map[nums[i]].clear()        # clear this level

            min_jumps += 1

        return -1
            

File: 1447-jump-game-iv/jump-game-iv.java
class Solution {
    public int minJumps(int[] arr) {
        // idx -> idxes
        Map<Integer, List<Integer>> idxMap = new HashMap<>();
        for (int i = 0; i < arr.length; i++) {
            idxMap.computeIfAbsent(arr[i], k -> new ArrayList<>()).add(i);
        }
        
        // bfs
        boolean[] visited = new boolean[arr.length];
        Deque<Integer> queue = new ArrayDeque<>();
        visited[0] = true;
        queue.offerLast(0);
        int steps = 0;
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            for (int i = 0; i < levelSize; i++) {
                int curIdx = queue.pollFirst();
                if (curIdx == arr.length - 1) return steps;

                if (idxMap.containsKey(arr[curIdx])) {
                    for (int nbr : idxMap.get(arr[curIdx])) {
                        if (!visited[nbr]) {
                            visited[nbr] = true;
                            queue.offerLast(nbr);
                        }
                    }
                    idxMap.remove(arr[curIdx]);
                }

                if (curIdx + 1 < arr.length && !visited[curIdx + 1]) {
                    visited[curIdx + 1] = true;
                    queue.offerLast(curIdx + 1);
                }

                if (curIdx - 1 >= 0 && !visited[curIdx - 1]) {
                    visited[curIdx - 1] = true;
                    queue.offerLast(curIdx - 1);
                }
            }
            steps += 1;
        }

        return -1;
    }
}